### PR TITLE
Add Stdlib.Ref

### PR DESCRIPTION
**Pull request by nojb**
After writing `with_ref` and `add_to_list` for Nth time, I wondered why we don't have them in the stdlib.

What do you think?

### CODE REVIEW

**Comment by xvw** [comment]
I don't use a lot of `ref` but with `incr` and `add_to_list` it seems very opiniated. What do you think about adding `val put : 'a ref -> ('a -> 'a) -> unit`?

**Comment by nojb** [comment]
> I don't use a lot of `ref` but with `incr` and `add_to_list` it seems very opiniated. What do you think about adding `val put : 'a ref -> ('a -> 'a) -> unit`?

Sure, let's see if there is a consensus about the module itself before discussing new functions though. Note that `incr` and `decr` are part of `Stdlib` already.

File: stdlib/ref.ml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+(*   Copyright 2020 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+type 'a t = 'a ref = {mutable contents : 'a}
```
Comment (rizo) [inline stdlib/ref.ml:16]: Would `[@@unboxed]` be useful here?

CC @c-cube

Comment (rizo) [reply stdlib/ref.ml:16]: "unboxable record types are the _immutable_ record types that have only one field" ([ref](https://caml.inria.fr/pub/docs/manual-ocaml/intfc.html#ss:c-tuples-and-records)) – I guess not.

File: stdlib/ref.mli [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+(** References.
+
+    References are mutable indirection cells containing a value of a given type.
+
+    @since 4.12 *)
+
+type 'a t = 'a ref = {mutable contents : 'a}
+(** The type of references of type ['a]. *)
+
+val ref : 'a -> 'a t
```
Comment (dbuenzli) [inline stdlib/ref.mli:25]: I think this should be `create`. Note that we have `Atomic.make` though I personally think it's a mistake that should be corrected before the module goes into the world. (There should be a discussion with @gasche about that somewhere but I fail to find it on the `Atomic` module PR).

File: stdlib/ref.mli [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+val set : 'a t -> 'a -> unit
+(** [set r x] sets the value stored in [r] to [x]. *)
+
+val with_ref : 'a t -> 'a -> (unit -> 'b) -> 'b
+(** [with_ref r x f] sets [r] to [x], and returns the result of [f], after
+    restoring [r] to its previous value (even if an exception is raised by
+    [f]. *)
+
+val equal : 'a t -> 'a t -> bool
```
Comment (dbuenzli) [inline stdlib/ref.mli:39]: There's a long discussion [here](https://github.com/ocaml/ocaml/pull/9080) about giving a name to physical equality. I'm a bit unsure if for clarity I would not prefer to have that name be used there aswell.

File: stdlib/ref.mli [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+(** [equal r r'] is [r == r']. *)
+
+val incr : int t -> unit
+(** [incr r] is [set r (get r + 1)]. *)
+
+val decr : int t -> unit
+(** [decr r] is [set r (get r - 1)]. *)
+
+val add_to_list : 'a list t -> 'a -> unit
```
Comment (rgrinberg) [inline stdlib/ref.mli:48]: Would a name like `cons_to_list` make more sense here? It's not entirely clear what `add` is for a list.

File: stdlib/ref.mli [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+(** [add_to_list r x] is [set r (x :: get r)]. *)
```
Comment (dbuenzli) [inline stdlib/ref.mli:49]: If I read `Ref.add_to_list l v` I would rather think a reference is being added to a list in someway. I think if you really want that function (I personally think it's better not to have too much convenience on references) you should find a better name.

**Comment by nojb** [comment]
Closing this, there is not enough motivation for this at the moment.

**Pull request closed without merging**